File: src/audio_stegano_app/lib/core/stego/capacity_exceeded_exception.dart
/// Thrown when a payload needs more LSB bits than the cover audio capacity.
class CapacityExceededException implements Exception {
  final int neededBits;
  final int availableBits;

  const CapacityExceededException({
    required this.neededBits,
    required this.availableBits,
  });

  /// Matches engine messages: `Message too long: needs N bits, capacity M`.
  static CapacityExceededException? tryParse(Object error) {
    if (error is CapacityExceededException) return error;
    final text = error.toString();
    final match = RegExp(
      r'needs\s+(\d+)\s+bits.*?capacity\s+(\d+)',
      caseSensitive: false,
    ).firstMatch(text);
    if (match == null) return null;
    final needed = int.tryParse(match.group(1)!);
    final available = int.tryParse(match.group(2)!);
    if (needed == null || available == null) return null;
    return CapacityExceededException(
      neededBits: needed,
      availableBits: available,
    );
  }

  @override
  String toString() =>
      'Message too long: needs $neededBits bits, capacity $availableBits';
}

File: src/audio_stegano_desktop/src/AudioStegano.Core/Stego/CapacityExceededException.cs
namespace AudioStegano.Core.Stego;

/// <summary>Payload needs more LSB bits than the cover mono sample capacity.</summary>
public sealed class CapacityExceededException : Exception
{
    public int NeededBits { get; }
    public int AvailableBits { get; }

    public CapacityExceededException(int neededBits, int availableBits)
        : base($"Message too long: needs {neededBits} bits, capacity {availableBits}")
    {
        NeededBits = neededBits;
        AvailableBits = availableBits;
    }

    /// <summary>Parse engine messages: Message too long: needs N bits, capacity M.</summary>
    public static bool TryParse(string? message, out CapacityExceededException? exception)
    {
        exception = null;
        if (string.IsNullOrWhiteSpace(message)) return false;
        var match = System.Text.RegularExpressions.Regex.Match(
            message,
            @"needs\s+(\d+)\s+bits.*?capacity\s+(\d+)",
            System.Text.RegularExpressions.RegexOptions.IgnoreCase);
        if (!match.Success) return false;
        if (!int.TryParse(match.Groups[1].Value, out var needed)) return false;
        if (!int.TryParse(match.Groups[2].Value, out var available)) return false;
        exception = new CapacityExceededException(needed, available);
        return true;
    }
}
